pkg/api/transport_custom_test.go

```diff
@@ -0,0 +1,258 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package api
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/elastic/cloud-sdk-go/pkg/api/mock"
+	"github.com/elastic/cloud-sdk-go/pkg/multierror"
+)
+
+func TestNewCustomTransport(t *testing.T) {
+	type args struct {
+		cfg CustomTransportCfg
+	}
+	tests := []struct {
+		name string
+		args args
+		want *CustomTransport
+		err  error
+	}{
+		{
+			name: "fails creating a custom transport due to invalid config",
+			args: args{cfg: CustomTransportCfg{
+				Verbose: true,
+			}},
+			err: multierror.NewPrefixed("invalid custom transport settings",
+				errors.New("roundtripper cannot be nil"),
+				errors.New("verbose set to true, but no writer has been set"),
+			),
+		},
+		{
+			name: "creates a new transport with all defaults",
+			args: args{cfg: CustomTransportCfg{
+				RoundTripper: http.DefaultTransport,
+			}},
+			want: &CustomTransport{
+				rt:      http.DefaultTransport,
+				agent:   DefaultUserAgent,
+				backoff: defaultBackoff,
+			},
+		},
+		{
+			name: "creates a new transport with custom configuration",
+			args: args{cfg: CustomTransportCfg{
+				RoundTripper: http.DefaultTransport,
+				UserAgent:    "someagent",
+				Retries:      5,
+				Backoff:      time.Millisecond,
+				Verbose:      true,
+				Writer:       ioutil.Discard,
+			}},
+			want: &CustomTransport{
+				rt:      http.DefaultTransport,
+				agent:   "someagent",
+				backoff: time.Millisecond,
+				retries: 5,
+				verbose: true,
+				writer:  ioutil.Discard,
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := NewCustomTransport(tt.args.cfg)
+			assert.Equal(t, tt.err, err)
+			assert.Equal(t, tt.want, got)
+		})
+	}
+}
+
+// nolint
```

Comment:
Just out of curiosity, what was the linting error you got here?

Reply:
oh, it was complaining that I wasn't closing `http.Request.Body` 😂 , so I just unlinted it